### PR TITLE
Feature/setup filter form

### PR DESCRIPTION
**Pull request by AdityaKhatri**

### CODE REVIEW

File: src/Root/App/index.tsx [hunk 1]
```diff
@@ -1,14 +1,24 @@
...
 import '../../../node_modules/mapbox-gl/dist/mapbox-gl.css';
 import './styles.css';
 
 import Home from '../../views/Home';
 
 function App() {
+    const requestContextValue = {
+        transformUrl: (d: string) => d,
+        transformOptions: (_: string, options) => processOptions(options),
+    };
```
Comment (tnagorra) [inline src/Root/App/index.tsx:16]: Please memoize requestContextValue.

File: src/Root/App/index.tsx [hunk 2]
```diff
@@ -1,14 +1,24 @@
...
 import '@togglecorp/toggle-ui/build/index.css';
 import '../../../node_modules/mapbox-gl/dist/mapbox-gl.css';
 import './styles.css';
 
 import Home from '../../views/Home';
 
 function App() {
+    const requestContextValue = {
+        transformUrl: (d: string) => d,
+        transformOptions: (_: string, options) => processOptions(options),
```
Comment (tnagorra) [inline src/Root/App/index.tsx:15]: Please specify type for options argument

File: src/Root/App/styles.css [hunk 3]
```diff
@@ -31,12 +31,15 @@
...
-    --color-primary: #354052;
-    --color-accent: #008dca;
+    --color-primary: #730fc3;
+    --color-accent: #6558f5;
     --color-danger: #e04656;
     --color-success: #26c681;
     --color-warning: #efe045;
 
+    --tui-color-primary: #730fc3;
+    --tui-color-accent: #6558f5;
```
Comment (tnagorra) [inline src/Root/App/styles.css:41]: Please re-use previously defined variable

```suggestion
    --tui-color-primary: var(--color-primary);
    --tui-color-accent: var(--color-accent);
```

File: src/Root/App/styles.css [hunk 4]
```diff
@@ -51,6 +54,7 @@
     /* COLOR GENERAL */
     --color-foreground: #ffffff;
     --color-background: #f0f3f7;
+    --color-background-alternate: #dfe6ed;
```
Comment (tnagorra) [inline src/Root/App/styles.css:57]: Please add comment how this is different from `color-background`

File: src/components/TagInput/index.tsx [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+                        <IoClose />
+                    )}
+                />
+            )}
+            variant={variant}
+        >
+            {label}
+        </TagView>
+    );
+}
```
Comment (tnagorra) [inline src/components/TagInput/index.tsx:60]: Isn't this already a separate component?

File: src/components/TagInput/index.tsx [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+    value?: string[];
+    label: string;
+    name: N;
+    variant?: TagVariant;
+    onChange?: (newVal: string[], name: N) => void;
+    disabled?: boolean;
+    readOnly?: boolean;
+}
+
+const keySelector = (d: string) => d;
```
Comment (tnagorra) [inline src/components/TagInput/index.tsx:73]: Please move keySelector before Props definition for TagInput

File: src/components/TagInput/index.tsx [hunk 7]
```diff
@@ -0,0 +1,208 @@
...
+                />
+            )}
+            variant={variant}
+        >
+            {label}
+        </TagView>
+    );
+}
+
+interface Props<N> {
```
Comment (tnagorra) [inline src/components/TagInput/index.tsx:62]:
```suggestion
interface Props<N extends string> {
```

File: src/components/TagInput/index.tsx [hunk 8]
```diff
@@ -0,0 +1,208 @@
...
+    disabled?: boolean;
+    readOnly?: boolean;
+}
+
+const keySelector = (d: string) => d;
+
+function TagInput<N extends string>(props: Props<N>) {
+    const {
+        className,
+        value = [],
```
Comment (tnagorra) [inline src/components/TagInput/index.tsx:78]: We should not create new object/array on render

File: src/components/TagInput/index.tsx [hunk 9]
```diff
@@ -0,0 +1,208 @@
...
+            variant,
+            disabled,
+            readOnly,
+        }),
+        [handleTagRemove, variant, readOnly, disabled],
+    );
+
+    const handleNewTagValueAdd = useCallback((newVal) => {
+        setNewTagValue(newVal);
+    }, []);
```
Comment (tnagorra) [inline src/components/TagInput/index.tsx:133]:
```suggestion
    const handleNewTagValueAdd = setNewTagValue;
```

File: src/components/TagInput/index.tsx [hunk 10]
```diff
@@ -0,0 +1,208 @@
...
+                                    childrenClassName={styles.children}
+                                    disabled={disabled}
+                                    icons={(
+                                        <IoClose />
+                                    )}
+                                />
+                            </>
+                        ) : (
+                            <Button
+                                name="remove"
```
Comment (tnagorra) [inline src/components/TagInput/index.tsx:181]: Did you mean name="add"?

File: src/hooks/useBooleanState.ts [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+
+    const setFalse = React.useCallback(() => {
+        setValue(false);
+    }, [setValue]);
+
+    const toggleFn = React.useCallback(() => {
+        setValue((oldValue) => !oldValue);
+    }, [setValue]);
+
+    return [value, setTrue, setFalse, setValue, toggleFn];
```
Comment (tnagorra) [inline src/hooks/useBooleanState.ts:29]:
```suggestion
    return [value, setTrue, setFalse, setValue, toggleFn] as const;
```

File: src/views/Home/Article/index.tsx [hunk 12]
```diff
@@ -0,0 +1,110 @@
...
+} from 'react-icons/io5';
+import {
+    DateTime,
+    Button,
+} from '@togglecorp/toggle-ui';
+
+import TagInput from '#components/TagInput';
+import { Article as ArticleProps } from '..';
+
+import styles from './styles.css';
```
Comment (tnagorra) [inline src/views/Home/Article/index.tsx:15]:
```suggestion
import TagInput from '#components/TagInput';

import { Article as ArticleProps } from '..';
import styles from './styles.css';
```

File: src/views/Home/Article/index.tsx [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+    IoClose,
+    IoCheckmarkSharp,
+} from 'react-icons/io5';
+import {
+    DateTime,
+    Button,
+} from '@togglecorp/toggle-ui';
+
+import TagInput from '#components/TagInput';
+import { Article as ArticleProps } from '..';
```
Comment (tnagorra) [inline src/views/Home/Article/index.tsx:13]:
```suggestion
import type { Article as ArticleProps } from '..';
```

File: src/views/Home/Article/index.tsx [hunk 14]
```diff
@@ -0,0 +1,110 @@
...
+                    variant="danger"
+                />
+                <Button
+                    name="relevant"
+                    className={styles.button}
+                    onClick={handleMarkAsRelevant}
+                    title="Mark as Relevant"
+                    icons={(
+                        <IoCheckmarkSharp />
+                    )}
```
Comment (tnagorra) [inline src/views/Home/Article/index.tsx:102]: Please pass icon as children when there is no text.

File: src/views/Home/Filters/index.tsx [hunk 15]
```diff
@@ -0,0 +1,263 @@
...
+        includeAllKeywords: [],
+        includeSomeKeywords: [],
+        includeNoKeywords: [],
+    }),
+};
+
+const defaultFormValues: FormType = {
+    search: '',
+    fuzzySearch: true,
+    includeAllKeywords: ['1st tag', '2nd tag'],
```
Comment (tnagorra) [inline src/views/Home/Filters/index.tsx:118]: Why did we add these as default values? Are they temporary?

File: src/views/Home/Filters/index.tsx [hunk 16]
```diff
@@ -0,0 +1,263 @@
...
+};
+
+interface Props {
+    className?: string;
+    onSearchClick: (params: FormType) => void;
+}
+
+function Home(props: Props) {
```
Comment (tnagorra) [inline src/views/Home/Filters/index.tsx:126]:
```suggestion
function HomeFilters(props: Props) {
```

File: src/views/Home/Filters/index.tsx [hunk 17]
```diff
@@ -0,0 +1,263 @@
...
+    const handleSubmit = useCallback((finalValues: FormType) => {
+        onSearchClick(finalValues);
+    }, [onSearchClick]);
+
+    return (
+        <form
+            className={_cs(className, styles.filters)}
+            onSubmit={createSubmitHandler(validate, onErrorSet, handleSubmit)}
+        >
+            <h1 className={styles.heading}>Explore MMP</h1>
```
Comment (tnagorra) [inline src/views/Home/Filters/index.tsx:155]:
```suggestion
            <h1 className={styles.heading}>
                Explore MMP
            </h1>
```

File: src/views/Home/Filters/index.tsx [hunk 18]
```diff
@@ -0,0 +1,263 @@
...
+                label="Include No Keywords"
+                value={value.includeNoKeywords}
+                variant="danger"
+                onChange={onValueChange}
+            />
+            <div className={styles.footerContainer}>
+                <Button
+                    className={styles.button}
+                    name="reset"
+                    onClick={handleFiltersReset}
```
Comment (tnagorra) [inline src/views/Home/Filters/index.tsx:245]: Can we not disable the reset button when there is no change on filters?

File: src/views/Home/index.tsx [hunk 19]
```diff
@@ -1,18 +1,136 @@
...
 interface Props {
     className?: string;
 }
 
 function Home(props: Props) {
     const { className } = props;
+    const [activePage, setActivePage] = useState(1);
+    const [pageSize, setPageSize] = useState(10);
+    const [totalCount, setTotalCount] = useState(0);
+    const [searchBody, setSearchBody] = useState(undefined);
```
Comment (tnagorra) [inline src/views/Home/index.tsx:74]: Let's move this to request context.
The type for searchBody is not defined!

File: src/views/Home/index.tsx [hunk 20]
```diff
@@ -1,18 +1,136 @@
...
+        url: 'https://abc.com',
+        skip: isNotDefined(searchBody),
+        query,
+        onSuccess: (response) => {
+            setTotalCount(response?.total);
+        },
+        mockResponse: fakeData,
+    });
+
+    const articleRendererParams = useCallback((_, data) => ({
```
Comment (tnagorra) [inline src/views/Home/index.tsx:94]: Please define type for data

File: src/views/Home/index.tsx [hunk 21]
```diff
@@ -1,18 +1,136 @@
...
+        ...data,
+    }), []);
+
+    const handleSearch = useCallback((searchParams) => {
+        setSearchBody(searchParams);
+    }, []);
```
Comment (tnagorra) [inline src/views/Home/index.tsx:100]:
```suggestion
    const handleSearch = setSearchBody;
```

File: src/views/Home/index.tsx [hunk 22]
```diff
@@ -1,18 +1,136 @@
...
         <div className={_cs(className, styles.home)}>
-            <h1>MMP Dashboard</h1>
+            <Filters
+                className={styles.leftContainer}
+                onSearchClick={handleSearch}
+            />
+            <div className={styles.rightContainer}>
+                {pending && (<PendingMessage />)}
+                <div className={styles.listContainer}>
+                    {articlesResponse?.results ? (
```
Comment (tnagorra) [inline src/views/Home/index.tsx:111]: Please also check if length is greater than zero!

File: src/views/Home/index.tsx [hunk 23]
```diff
@@ -1,18 +1,136 @@
...
+                        <List
+                            data={articlesResponse.results}
+                            renderer={Article}
+                            rendererParams={articleRendererParams}
+                            keySelector={articleKeySelector}
+                        />
+                    ) : (
+                        <h2 className={styles.emptyMessage}>
+                            There are no results available.
+                        </h2>
```
Comment (tnagorra) [inline src/views/Home/index.tsx:121]: This is semantically not correct!

**Review by tnagorra — CHANGES_REQUESTED** [review]

File: src/components/TagInput/index.tsx (hunk 5 above)
Comment (AdityaKhatri) [reply src/components/TagInput/index.tsx:60]: No, It has been included as TagView

File: package.json [hunk 24]
```diff
@@ -14,14 +14,13 @@
     "scripts": {
         "start": "webpack-dev-server --mode development --config config/webpack.dev.js --env.NODE_ENV=development --history-api-fallback --config-register @babel/register --color --disable-host-check",
         "build:raw": "webpack --mode production --config config/webpack.prod.js --env.NODE_ENV=production --config-register @babel/register",
-        "build": "yarn run generate && yarn run typecheck && yarn run build:raw",
-        "build:unsafe": "mkdir -p generated && touch generated/types.tsx && yarn run build:raw",
+        "build": "yarn run build:raw",
```
Comment (tnagorra) [inline package.json:17]: Why not just rename `build:raw` to `build`

File: package.json [hunk 25]
```diff
@@ -14,14 +14,13 @@
...
         "serve": "http-server ./build/ -p 3081 -g -c-1",
         "test": "jest",
         "test:watch": "jest --watch",
         "test:coverage": "jest --coverage",
-        "typecheck": "tsc",
-        "lint": "eslint ./src"
+        "lint": "eslint ./src",
```
Comment (tnagorra) [inline package.json:22]: Why remove typecheck?

**Review by tnagorra — CHANGES_REQUESTED** [review]

File: src/views/Login/index.tsx [hunk 26]
```diff
@@ -0,0 +1,116 @@
...
+    } = useLazyRequest({
+        method: 'POST',
+        url: '/api/login',
+        body: (ctx) => ctx,
+    });
+
+    const handleSubmit = (finalValue: FormType) => {
+        const completeValue = finalValue as LoginFormFields;
+        login(finalValue);
+        console.warn('i am here', completeValue, setUser);
```
Comment (tnagorra) [inline src/views/Login/index.tsx:65]:
```suggestion
```

**Review by tnagorra — CHANGES_REQUESTED** [review]

File: src/Root/App/Multiplexer/index.tsx [hunk 27]
```diff
@@ -0,0 +1,102 @@
...
+    }
+
+    if (waiting) {
+        return (
+            <div className={_cs(className, styles.multiplexer)}>
+                <PendingMessage message="Checking user session..." />
+            </div>
+        );
+    }
+     */
```
[the remark this answers is not in the document]
Comment (AdityaKhatri) [reply src/Root/App/Multiplexer/index.tsx:59]: This is done to skip auth.
As soon as we setup auth, we'll have this back.

File: src/views/Home/Filters/index.tsx (hunk 18 above)
Comment (AdityaKhatri) [reply src/views/Home/Filters/index.tsx:250]: We can, but chose not to :D

Comment (tnagorra) [reply src/views/Home/Filters/index.tsx:250]: Please choose to do it.

File: src/views/Home/Filters/index.tsx [hunk 28]
```diff
@@ -0,0 +1,271 @@
...
+    const handleFiltersReset = useCallback(() => {
+        onValueSet(defaultFormValues);
+        onSearchClick(defaultFormValues);
+    }, [onValueSet, onSearchClick]);
+
+    const handleSubmit = useCallback((finalValues: FormType) => {
+        onSearchClick(finalValues);
+    }, [onSearchClick]);
+
+    const disableReset = value === defaultFormValues;
```
Comment (tnagorra) [inline src/views/Home/Filters/index.tsx:153]:
```suggestion
    const resetDisabled = value === defaultFormValues;
```

**Review by tnagorra — CHANGES_REQUESTED** [review]

**Review by tnagorra — APPROVED** [review]

**Pull request merged**